### PR TITLE
SLING-11824 : Deadlock during repository restart

### PR DESCRIPTION
**Pull request by cziegeler**

### CODE REVIEW

File: src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java [hunk 1]
```diff
@@ -116,6 +116,9 @@ public class SlingServletContext implements ServletContext, ServletContextListen
 
     private volatile String configuredServerInfo;
 
+    // counter to synchronize the init and destroy methods
+    private volatile long initCounter;
```
Comment (jsedding) [inline src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java:120]: nit: wouldn't an int suffice? IIUC we don't care about an overflow, we only care that it's different.

File: src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java [hunk 2]
```diff
@@ -199,37 +202,78 @@ private void setServerInfo() {
...
-    public void contextDestroyed(final ServletContextEvent sce) {
-        synchronized ( this ) {
-            this.servletContext = null;
-            this.setServerInfo();
-            if ( this.registration != null ) {
-                this.registration.unregister();
-                this.registration = null;    
-            }
+    private void runAsync(final Runnable r) {
+        final Thread thread = new Thread(r, "SlingServletContext registration");
```
Comment (jsedding) [inline src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java:206]: I think having different thread names for register and unregister might be useful.

**Review by jsedding — APPROVED** [review]
Looks good to me, both the register and unregister calls are now outside of the synchronized blocks. I added two comments for your consideration.

Do you think it would be feasible to add an automated test?

File: src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java (hunk 1 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/engine/impl/helper/SlingServletContext.java:120]: Sure, but it doesn't really matter - so I stick with long :)

**Comment by cziegeler** [comment]
I've added some basic tests

**Pull request merged**